block-producer/src/test_utils/store.rs

```diff
@@ -3,63 +3,91 @@ use std::collections::BTreeSet;
 use async_trait::async_trait;
 use miden_objects::{
     crypto::merkle::{Mmr, SimpleSmt, Smt, ValuePath},
-    notes::Nullifier,
-    BlockHeader, ACCOUNT_TREE_DEPTH, EMPTY_WORD, ONE, ZERO,
+    notes::{NoteEnvelope, Nullifier},
+    BlockHeader, ACCOUNT_TREE_DEPTH, BLOCK_OUTPUT_NOTES_TREE_DEPTH, EMPTY_WORD, ONE, ZERO,
 };
 
 use super::*;
 use crate::{
+    batch_builder::TransactionBatch,
     block::{AccountWitness, Block, BlockInputs},
     store::{
         ApplyBlock, ApplyBlockError, BlockInputsError, Store, TransactionInputs, TxInputsError,
     },
+    test_utils::block::{note_created_smt_from_batches, note_created_smt_from_envelopes},
     ProvenTransaction,
 };
 
 /// Builds a [`MockStoreSuccess`]
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct MockStoreSuccessBuilder {
     accounts: Option<SimpleSmt<ACCOUNT_TREE_DEPTH>>,
-    produced_nullifiers: Option<Smt>,
+    notes: Option<SimpleSmt<BLOCK_OUTPUT_NOTES_TREE_DEPTH>>,
+    produced_nullifiers: Option<BTreeSet<Digest>>,
     chain_mmr: Option<Mmr>,
     block_num: Option<u32>,
 }
 
 impl MockStoreSuccessBuilder {
-    /// FIXME: the store always needs to be properly initialized with initial accounts
-    /// see https://github.com/0xPolygonMiden/miden-node/issues/79
-    pub fn new() -> Self {
-        Self::default()
+    pub fn from_batches<'a>(batches: impl Iterator<Item = &'a TransactionBatch>) -> Self {
+        let batches: Vec<_> = batches.collect();
+
+        let accounts_smt = {
+            let accounts = batches
+                .iter()
+                .cloned()
+                .flat_map(TransactionBatch::account_initial_states)
+                .map(|(account_id, hash)| (account_id.into(), hash.into()));
+            SimpleSmt::<ACCOUNT_TREE_DEPTH>::with_leaves(accounts).unwrap()
+        };
+
+        let created_notes = note_created_smt_from_batches(batches.iter().cloned());
+        let produced_nullifiers = batches
+            .iter()
+            .cloned()
+            .flat_map(TransactionBatch::produced_nullifiers)
+            .map(|nullifier| nullifier.inner())
+            .collect();
```

Comment:
This should be removed, and fixes the test (as you mentioned offline).

The problem is that you're initializing the store with the "new" nullifiers that the transaction wants to update. That is, when a batch produces nullifiers (i.e. sets their value to `[block_num, 0, 0, 0]` in the store), they should have a value of `[0, 0, 0, 0]` in the store. Thus, in this function, we should *not* take the produced nullifiers and add them to the store - the block kernel will be doing that.

The error `BlockProverFailed(ProgramExecutionFailed(MerkleStoreLookupFailed(RootNotInStore(...))))` is doesn't give much insight into why this fails (maybe we can fix that). When we initialize the Merkle store from the block witness, we basically assume that the value is `0` in the store ([here](https://github.com/0xPolygonMiden/miden-node/blob/f323ffa15d240388a41bc68953b5995ba47ac68f/block-producer/src/block_builder/prover/block_witness.rs#L292)). But since you set the value to something else here, the Merkle store root becomes inconsistent with the actual store's root (due to the fact that 2 nullifiers are set in the Merkle store, and not set in the actual store). So when we try to `smt::set` the new nullifier [here](https://github.com/0xPolygonMiden/miden-node/blob/f323ffa15d240388a41bc68953b5995ba47ac68f/block-producer/src/block_builder/prover/mod.rs#L127), we provide the incorrect Merkle store root, which the VM complains about.

Reply:
That was my main concern! Thank you! I've fixed this and CI is expected to pass now. Please, make a review if you have a chance!